reconcile/vault_replication.py

```diff
@@ -0,0 +1,255 @@
+import logging
+import re
+
+from reconcile.utils.vault import VaultClient, _VaultClient, SecretNotFound
+from reconcile.gql_definitions.jenkins_configs import jenkins_configs
+from reconcile.gql_definitions.jenkins_configs.jenkins_configs import (
+    JenkinsConfigV1_JenkinsConfigV1,
+)
+
+from reconcile.gql_definitions.vault_policies import vault_policies
+from reconcile.gql_definitions.vault_instances import vault_instances
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultInstanceAuthApproleV1,
+    VaultInstanceV1,
+    VaultReplicationConfigV1_VaultInstanceV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    VaultReplicationConfigV1_VaultInstanceV1,
+)
+
+from reconcile.utils import gql
+from typing import List, cast, Optional, Union
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+class VaultInvalidPaths(Exception):
+    pass
+
+
+class VaultInvalidAuthMethod(Exception):
+    pass
+
+
+def deep_copy_versions(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    current_dest_version: int,
+    current_source_version: int,
+    path: str,
+):
+    for version in range(current_dest_version + 1, current_source_version + 1):
+        secret_dict = {"path": path, "version": version}
+        secret, src_version = source_vault.read_all_with_version(secret_dict)
+        write_dict = {"path": path, "data": secret}
+        logging.info(["replicate_vault_secret", src_version, path])
+        if not dry_run:
+            dest_vault.write(write_dict)
+
+
+def copy_vault_secret(
+    dry_run: bool, source_vault: _VaultClient, dest_vault: _VaultClient, path: str
+) -> None:
+
+    secret_dict = {"path": path, "version": "LATEST"}
+    _, version = source_vault.read_all_with_version(secret_dict)
+
+    try:
+        _, dest_version = dest_vault.read_all_with_version(secret_dict)
+        if dest_version is None and version is None:
+            # v1 secrets don't have version
+            secret, _ = source_vault.read_all_with_version(secret_dict)
+            write_dict = {"path": path, "data": secret}
+            logging.info(["replicate_vault_secret", path])
+            if not dry_run:
+                dest_vault.write(write_dict)
+        elif dest_version < version:
+            deep_copy_versions(
+                dry_run=dry_run,
+                source_vault=source_vault,
+                dest_vault=dest_vault,
+                current_dest_version=dest_version,
+                current_source_version=version,
+                path=path,
+            )
+        else:
+            logging.info(["replicate_vault_secret", dest_version, version, path])
```

Comment:
is this log line necessary? This is going to cover the case dest_version == version, in which we don't replicate anything, isn't it?

Reply:
Removed the else condition